site/content/features/typepatterns.md

```diff
@@ -525,6 +505,70 @@ public class Main {
 {{< /sandboxsource >}}
 {{< /sandbox >}}
 
+## Enum Constants
+
+When the selector of a `switch` has `enum` type, then the `case` labels are constants in the enumeration:
+
+```
+System.Logger.Level level = ...;
+switch (level) {
+   case ERROR: ...; // i.e. System.Logger.Level.ERROR
+   ...
+}
+```
+
+Note also that the labels are not “qualified”—that is, you omit the enum type name in a `case` label.
+
+Now consider a more complex case—a sealed hierarchy in which some subtypes are enumerations:
+
+```
+JSONPrimitive p = ...;
+switch (p) {
+   case JSONNumber n -> ...;
+   case JSONString s -> ...;
+   case JSONNull.INSTANCE -> ...; // Can't be case INSTANCE
+   case JSONBoolean.FALSE -> ...; 
+   case JSONBoolean.TRUE -> ...;
+}
+```
+
+In this situation, you can use case labels with `enum` constants. However, they need to be qualified, since the compiler cannot deduce the `enum` type from the selector.
+
+With `enum` constants (but not with numeric or `String` constants), the rule that the selector type must match the constant type has been relaxed. The selector type can be any supertype.
+
+{{< sandbox version=java21 preview="true" mainclass="Main" >}}{{< sandboxsource "Main.java" >}}
```

Comment:
I get the following compiler error:

    Main.java:9: error: ';' expected
          }

After fixing this definition of `JSONValue` is missing.

Also `preview` not required here.

Reply:
Fixing right now

Reply:
Still compile errors here:

```
Main.java:4: error: cannot find symbol
      int result = switch (value) {
                           ^
  symbol:   variable value
  location: class Main
Main.java:5: error: cannot find symbol
         case JSONNumber n -> n.intValue();
                               ^
  symbol:   method intValue()
  location: variable n of type JSONNumber
Main.java:6: error: duplicate unconditional pattern
         case JSONString s -> Integer.parseInt(p.value);
              ^
Main.java:6: error: cannot find symbol
         case JSONString s -> Integer.parseInt(p.value);
                                                ^
  symbol:   variable value
  location: variable p of type JSONPrimitive
```

Reply:
Sorry, just fixed them.

Reply:
Well, different compile errors now...

```
Main.java:5: error: cannot find symbol
         case JSONNumber n -> n.intValue();
                               ^
  symbol:   method intValue()
  location: variable n of type JSONNumber
Main.java:6: error: cannot find symbol
         case JSONString s -> Integer.parseInt(p.value);
                                               ^
  symbol:   variable p
  location: class Main
```

Reply:
Sorry, it runs now.